clients/src/main/java/org/apache/kafka/common/telemetry/internals/KafkaMetricsCollector.java

```diff
@@ -0,0 +1,341 @@
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.Measurable;
+import org.apache.kafka.common.metrics.MetricValueProvider;
+import org.apache.kafka.common.metrics.stats.Avg;
+import org.apache.kafka.common.metrics.stats.CumulativeCount;
+import org.apache.kafka.common.metrics.stats.CumulativeSum;
+import org.apache.kafka.common.metrics.stats.Frequencies;
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.common.metrics.stats.Meter;
+import org.apache.kafka.common.metrics.stats.Min;
+import org.apache.kafka.common.metrics.stats.Percentiles;
+import org.apache.kafka.common.metrics.stats.Rate;
+import org.apache.kafka.common.metrics.stats.SimpleRate;
+import org.apache.kafka.common.metrics.stats.WindowedCount;
+import org.apache.kafka.common.telemetry.internals.LastValueTracker.InstantAndValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Field;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * All metrics implement the {@link MetricValueProvider} interface. They are divided into
+ * two base types:
+ *
+ * <ol>
+ *     <li>{@link Gauge}</li>
+ *     <li>{@link Measurable}</li>
+ * </ol>
+ *
+ * {@link Gauge Gauges} can have any value but we only collect metrics with number values.
+ * {@link Measurable Measurables} are divided into simple types with single values
+ * ({@link Avg}, {@link CumulativeCount}, {@link Min}, {@link Max}, {@link Rate},
+ * {@link SimpleRate}, and {@link CumulativeSum}) and compound types ({@link Frequencies},
+ * {@link Meter}, and {@link Percentiles}).
+ *
+ * <p>
+ *
+ * We can safely assume that a {@link CumulativeCount count} always increases in steady state. It
+ * should be a bug if a count metric decreases.
+ *
+ * <p>
+ *
+ * Total and Sum are treated as a monotonically increasing counter. The javadocs for Total metric type
+ * say "An un-windowed cumulative total maintained over all time.". The standalone Total metrics in
+ * the codebase seem to be cumulative metrics that will always increase. The Total metric underlying
+ * Meter type is mostly a Total of a Count metric.
+ * We can assume that a Total metric always increases (but it is not guaranteed as the sample values might be both
+ * negative or positive).
+ * For now, Total is converted to CUMULATIVE_DOUBLE unless we find a valid counter-example.
+ *
+ * <p>
+ *
+ * The Sum as it is a sample sum which is not a cumulative metric. It is converted to GAUGE_DOUBLE.
+ *
+ * <p>
+ *
+ * The compound metrics are virtual metrics. They are composed of simple types or anonymous measurable types
+ * which are reported. A compound metric is never reported as-is.
+ *
+ * <p>
+ *
+ * A Meter metric is always created with and reported as 2 KafkaExporter metrics: a rate and a
```

Comment:
KafkaExporter refers to a class that doesn't exist in this codebase

Reply:
My bad, removed.